### PR TITLE
Migration to ndslice.algorithm and Mir

### LINKED ISSUE #20: migration to ndslice.algorithm and Mir
The common pattern for dcv is `aSlice.byElement.rangedFunction`. It has few performance issues:
1. It can not be vectroized
2. It requires additional computations because range interface

So, you may want to swtich from `dcv.core.algorithm` to `ndslice.algorithm`. 

The documentation for `ndslice.algorithm` can be found [here](docs.mir.dlang.io/latest/mir_ndslice_algorithm.html). `ndslice.algorithm` is currently available only in Mir.

So, switching to Mir is good option. It is providing recent and upcoming ndslice changes for both DMD and LDC. Also, mir will migrate to Phobos's ndslice after 2-3 DMD releases and will provide deprecation imports with aliasing. `ndslice,algorithm` is not the last module in ndslice package, also `ndslice.concatenation` will be added this year.

Please ask me questions if you have any.

### PR DESCRIPTION
**Pull request by ljubobratovicrelja**
Introducing following changes (should resolve #20):
- added Mir as dependency
- std.experimental.ndslice --> mir.ndslice
- slice.byElement.map/each/reduce/fold --> ndMap/ndEach/ndReduce/ndFold
- reformating of dcv.core.algorithm according to previous changes
- minor refactoring and optimization of library and example code related to dcv.core.algorithm changes

Eventually more algorithm optimizations will follow due to ndslice.algorithm. For this PR I didn't want to go too deep into changes described in #27, since those seemed to be too extensive and off-topic for this PR. So, we'll attend to that in future. 

Pinging @9il @DmitryOlshansky if you're willing to review.

### CODE REVIEW

File: source/dcv/core/algorithm.d [hunk 1]
```diff
@@ -36,252 +40,245 @@ enum NormType
...
-{
-    return range.findMinMax!">";
+    final switch (normType)
+    {
+    case NormType.INF:
+        return min.ndReduce!max(tensor).to!real;
+    case NormType.L1:
+        return T(0).ndReduce!((a, b) => abs(a + b))(tensor).to!real;
+    case NormType.L2:
+        return T(0).ndReduce!((a, b) => a + b ^^ 2)(tensor).to!real.sqrt;
```
Comment (9il) [inline source/dcv/core/algorithm.d]: `b * b` is better for optimization

File: source/dcv/core/algorithm.d [hunk 2]
```diff
@@ -36,252 +40,245 @@ enum NormType
...
+    tensor = Tensor of which the norm value is calculated.
+    normType = requested type of norm.
 
 Returns:
-    Minimum value of all elements in given range.
+    Calculated norm value as real.
 */
-ElementType!Range findMin(Range)(Range range) pure nothrow 
-        if (isForwardRange!Range && isNumeric!(ElementType!Range))
+@nogc pure nothrow real norm(Range, size_t N)(auto ref Slice!(N, Range) tensor, NormType normType)
```
Comment (9il) [inline source/dcv/core/algorithm.d]: why not `auto`?

File: source/dcv/core/algorithm.d [hunk 3]
```diff
@@ -36,252 +40,245 @@ enum NormType
...
-    Maximum value of all elements in given range.
-*/
-ElementType!Range findMax(Range)(Range range) pure nothrow 
-        if (isForwardRange!Range && isNumeric!(ElementType!Range))
-{
-    return range.findMinMax!">";
+    final switch (normType)
+    {
+    case NormType.INF:
+        return min.ndReduce!max(tensor).to!real;
```
Comment (9il) [inline source/dcv/core/algorithm.d]: why `.to!real`?

File: examples/filter/source/app.d [hunk 4]
```diff
@@ -58,10 +60,10 @@ int main(string[] args)
     auto medBlur = noisyImage.medianFilter(5);
 
     // scale values from 0 to 255 to preview gradient direction and magnitude
-    xgrads = xgrads.byElement.ranged(0, 255).array.sliced(xgrads.shape);
+    xgrads.ranged(0, 255);
     // Take absolute values and range them from 0 to 255, to preview edges
-    laplaceEdges = laplaceEdges.byElement.map!(a => fabs(a)).ranged(0, 255).array.sliced(laplaceEdges.shape);
-    logEdges = logEdges.byElement.map!(a => fabs(a)).ranged(0, 255).array.sliced(logEdges.shape);
+    laplaceEdges = laplaceEdges.ndMap!(a => fabs(a)).byElement.array.sliced(laplaceEdges.shape).ranged(0.0f, 255.0f);
```
Comment (9il) [inline examples/filter/source/app.d]: Faster variant:
`laplaceEdges.ndMap!(a => ranged(0.0f, 255.0f, fabs(a))).slice;`

You do not need `byElement.array.sliced` anymore

File: examples/filter/source/app.d [hunk 5]
```diff
@@ -58,10 +60,10 @@ int main(string[] args)
...
+    logEdges = logEdges.ndMap!(a => fabs(a)).byElement.array.sliced(logEdges.shape).ranged(0.0f, 255.0f);
```
Comment (9il) [inline examples/filter/source/app.d]: ditto

File: source/dcv/core/algorithm.d [hunk 6]
```diff
@@ -36,252 +40,245 @@ enum NormType
...
+    import std.conv : to;
 
-/// Find minimal value in the array.
-unittest
-{
-    auto arr = [1, 2, 3];
-    assert(arr.findMin == 1);
-}
+    version (LDC)
+        import ldc.intrinsics : sqrt = llvm_sqrt, abs = llvm_abs;
```
Comment (9il) [inline source/dcv/core/algorithm.d]: `fabs = llvm_fabs`

File: source/dcv/core/algorithm.d [hunk 7]
```diff
@@ -36,252 +40,245 @@ enum NormType
...
+    else
+        import std.math : sqrt, abs;
```
Comment (9il) [inline source/dcv/core/algorithm.d]: `fabs`

File: source/dcv/core/algorithm.d [hunk 8]
```diff
@@ -36,252 +40,245 @@ enum NormType
...
-ElementType!Range findMax(Range)(Range range) pure nothrow 
-        if (isForwardRange!Range && isNumeric!(ElementType!Range))
-{
-    return range.findMinMax!">";
+    final switch (normType)
+    {
+    case NormType.INF:
+        return min.ndReduce!max(tensor).to!real;
+    case NormType.L1:
+        return T(0).ndReduce!((a, b) => abs(a + b))(tensor).to!real;
```
Comment (9il) [inline source/dcv/core/algorithm.d]: `fabs`

File: source/dcv/core/algorithm.d (hunk 2 above)
Comment (9il) [inline source/dcv/core/algorithm.d]: `normL1`, `normL2`, `normInf` Looks better, imho

File: source/dcv/core/algorithm.d [hunk 9]
```diff
@@ -36,252 +40,245 @@ enum NormType
...
-    Returns the map result range that evaluates value scaling lazily.
 */
-auto ranged(Scalar, Range)(Range range, Scalar min = 0, Scalar max = 1)
-        if (isForwardRange!Range && isNumeric!(ElementType!Range)
-            && isNumeric!Scalar && isAssignable!(ElementType!Range, Scalar))
+@nogc auto ranged(Scalar, Range, size_t N)(auto ref Slice!(N, Range) tensor,
+        Scalar minValue = 0, Scalar maxValue = 1) if (isNumeric!Scalar)
 {
     import std.traits : isFloatingPoint;
+    import std.algorithm.comparison : min, max;
```
Comment (9il) [inline source/dcv/core/algorithm.d]: `llvm_maxnum` will be vectorized on recent LLVMs

**Comment by 9il** [comment]
`ndFold` should be used only for multiple functions. `ndReduce` is prefered and can be vectorized

**Comment by 9il** [comment]
@ljubobratovicrelja  Well done! I think it can be merged and comments can be referenced in a separate PRs

**Comment by ljubobratovicrelja** [comment]
> Well done! I think it can be merged and comments can be referenced in a separate PRs

@9il Thanks a ton! All comments are awesome and truly educational to me! :) Hope you don't mind, I'd like to address all comments in this PR - in any case, there'll be similar fixes throughout the library, so I'd like to fix these instances as example to myself and others.

**Pull request merged**